x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -461,7 +460,9 @@ private LogicalPlan resolveStats(Stats stats, List<Attribute> childrenOutput) {
             if (stats.expressionsResolved() == false) {
                 AttributeMap<Expression> resolved = new AttributeMap<>();
                 for (Expression e : groupings) {
-                    Attribute attr = Expressions.attribute(e);
+                    // In case of union types, the expression may contain yet-to-be resolved conversion functions, so do not perform a
+                    // type check here.
+                    Attribute attr = Expressions.attributeUnchecked(e);
```

Comment:
Oh no, forcing a reference attribute here is also wrong, because in case of truly unresolved expressions, calling `.datatype()` is wrong. I'll figure out a way around this.

Reply:
Thanks for the explanation - replace the Alias logic here; if it's an alias , return a ReferenceAttribute otherwise call toAttribute().
This gets rid of the attributeUnchecked() method